app/src/main/java/org/p2p/wallet/solana/SolanaNetworkObserver.kt

```diff
@@ -0,0 +1,133 @@
+) {
+
+    private val state = MutableStateFlow<SolanaNetworkState>(Idle)
+
+    private fun updateState(newState: SolanaNetworkState) {
+        Timber.tag(TAG).d("Updating the network state: $newState")
+        this.state.value = newState
+    }
+
+    fun getStateFlow(): Flow<SolanaNetworkState> = state
+
+    fun start() {
+        if (!observationFeatureToggle.isFeatureEnabled) {
+            Timber.tag(TAG).d("Solana network observation is disabled by feature toggle. Skipping the launch")
+            return
+        }
+
+        appScope.launch {
+            while (isActive) {
+                try {
+                    val samples = rpcSolanaRepository.getRecentPerformanceSamples(SAMPLE_COUNT)
+                    handleSamples(samples)
+                } catch (e: CancellationException) {
+                    Timber.w("Fetching recent performance samples cancelled")
+                } catch (e: Throwable) {
+                    Timber.e(e, "Error loading recent samples")
+                    showError()
+                }
+
+                /*
+                * Checking the network state every 10 seconds
+                * */
+                delay(REQUEST_REPEAT_TIME_IN_MS)
+            }
+        }
+    }
+
+    private fun handleSamples(samples: List<RecentPerformanceSample>) {
+        val transactions = samples.sumOf { it.numTransactions }
+        val periods = samples.sumOf { it.samplePeriodSecs }
+        val currentAverageTps = transactions / periods
+        Timber.tag(TAG).d("The average tps is: $currentAverageTps. Saving the value")
+
+        when (val oldState = state.value) {
+            is Idle,
+            is ShowError -> updateState(Online(currentAverageTps))
+            is Online -> calculateNegativePercent(oldState, currentAverageTps)
+        }
+    }
+
+    private fun calculateNegativePercent(
+        oldState: Online,
+        currentAverageTps: Int
+    ) {
+        val oldAverageTps = oldState.averageTps
+        val minAllowedTpsInPercent = percentFeatureToggle.value
+
+        /*
+         * Converting percent in actual TPS (Transactions Per Second) value
+         * */
+        val minAllowedTps = oldAverageTps * minAllowedTpsInPercent / 100
+
+        /**
+         * If the TPS less than allowed then we are showing error snackbar
+         * @example
+         * 2900 -> OLD TPS
+         * 1200 -> NEW TPS
+         * 70 percent == 2030 TPS
+         *
+         * Then 2030 TPS at LEAST allowed. If the [currentAverageTps] is less than that value - showing error
+         * */
+        if (currentAverageTps < minAllowedTps) {
+            showError()
+        } else {
+            updateState(Online(currentAverageTps))
+        }
+    }
+
+    /*
+    * Checking if user already saw the error message and we should show the message again or not
+    * according to the frequency value from RemoteConfig
+    * */
+    private fun showError() {
+        when (frequencyFeatureToggle.frequency) {
```

Comment:
i thought frequencyFeatureToggle -> is a frequency to fetch network state, not frequency to show error
do you mind changing the name by adding more context maybe?

Reply:
I can name it `errorFrequencyFeatureToggle`. 
Cannot come up with smth more

Reply:
showErrorFrequencyFeatureToggle / showErrorPeriodFeatureToggle?

Reply:
errorFrequencyFeatureToggle is okay too